native/src/main/java/io/ballerina/lib/data/xmldata/FromString.java

```diff
@@ -104,6 +109,38 @@ public static Object fromStringWithType(BString string, Type expType) {
         }
     }
 
+    private static Object stringToFiniteType(String value, FiniteType finiteType) {
+        return finiteType.getValueSpace().stream()
+                .filter(finiteValue -> !(convertToSingletonValue(value, finiteValue) instanceof BError))
+                .findFirst()
+                .orElseGet(() -> returnError(value, finiteType.toString()));
+    }
+
+    private static Object convertToSingletonValue(String str, Object singletonValue) {
+        String singletonStr = String.valueOf(singletonValue);
+        Type type = TypeUtils.getType(singletonValue);
+
+        if (singletonValue instanceof BDecimal decimalValue) {
+            BigDecimal bigDecimal = decimalValue.decimalValue();
+            if (bigDecimal.compareTo(new BigDecimal(str)) == 0) {
+                return fromStringWithType(StringUtils.fromString(str), type);
+            }
+            return returnError(str, singletonStr);
+        }
+
+        if (singletonValue instanceof Double doubleValue) {
+            if (doubleValue.compareTo(Double.valueOf(str)) == 0) {
+                return fromStringWithType(StringUtils.fromString(str), type);
+            }
```

Comment:
Instead of creating new double object you can use
```
if (doubleValue == Double.parseDouble(str)) {
    return fromStringWithType(StringUtils.fromString(str), type);
}

```

Reply:
In here we dont need to check the equality.
We need to check the numeric value equavility